tutorial/LFRic/background/LFRic_algorithm.md

```diff
@@ -0,0 +1,202 @@
+(defined as derived types, as explained in the [*LFRic kernel*
+section](LFRic_kernel.md). An example of such call can be found in
+the completed algorithm in [Example 1, solution of Part 1](
+../example1/solutions/part1/example1/example1_alg_mod.x90):
+
+```fortran
+    call invoke( setval_field_w0_kernel_type(field_w0, scalar_w0) )
+```
+
+The kernel objects must be made accessible to an algorithm through `use`
+statements, e.g.
+
+```fortran
+  use setval_field_w0_kernel_mod, only : setval_field_w0_kernel_type
+```
+
+The `invoke` calls to built-ins simply use the defined built-in names
+in PSyclone (see [built-ins naming scheme](
+https://psyclone.readthedocs.io/en/stable/dynamo0p3.html#naming-scheme)
+for more information), e.g. looking at the completed algorithm in
+[Example 2 solution](../example2/solutions/example2_alg_mod.x90):
+
+```fortran
+    call invoke( ...
+                 X_plus_Y(field_out_w0, field1_in_w0, field2_in_w0), &
+                 ... )
+```
+
+Built-ins do not need to be defined through the algorithm-level `use`
+statements as the code for the mathematical operations they perform is
+generated in the [PSy layer](LFRic_PSy.md).
+
+These examples of `invoke` calls illustrate the principle of operating on
+objects rather than directly on the object data in the algorithm layer.
+In this case it is passing the LFRic fields and not the field data as the
+arguments to kernel and built-in calls.
+
+### Grouping `invoke` calls and named `invoke`s
+
+It recommended to group as many kernel and built-in calls within a single
+`invoke` as possible when writing algorithm code to enable PSyclone to
+be effective in performing optimisations.
+
+As each `invoke` call generates a PSy-layer subroutine (see below for
+algorithm code processing), it can be easier to specify a [label (name)](
+https://psyclone.readthedocs.io/en/stable/algorithm_layer.html#named-invokes)
+for the `invoke` (and hence the subroutine) call to make it easier to search
+for in the generated code. For an example of a named invoke look into e.g.
+the completed algorithm in [Example 2 solution](
+../example2/solutions/example2_alg_mod.x90):
+
+```fortran
+    call invoke( name = "Builtins on W0 and W3 fields",              &
+                 ...
+                 X_plus_Y(field_out_w0, field1_in_w0, field2_in_w0), &
+                 ... )
+```
+
+### `use` statements and encapsulation
+
+As can be seen from the example algorithms
+[`../example1/example1_alg_mod.x90`](
+../example1/example1_alg_mod.x90) and [`../example2/example2_alg_mod.x90`](
+../example2/example2_alg_mod.x90), the `use` statements in algorithms
+mainly serve to access the LFRic infrastructure objects (e.g. mesh,
+function space, field).
+
+Unlike [kernels](LFRic_kernel.md), there is no inheritance from an
+abstract base type. Similar to kernels, however, the general
+encapsulation principle of keeping data and procedures `private`
+unless they are called by other objects still holds.
```

Comment:
This really just good general practice but it may not add anything to point that out.

Reply:
Thought so :)